### ISSUE TITLE
Vm instead of generator 

### ISSUE THREAD

**Issue by gedw99** [a-h/templ#103]
Templ is very nice but I can’t use it due to the golang generator. 

It would be nice if a golang vm thst parses and then generated the html at runtime was available.  

It wil be slower of course but provide a nicer developer experience for certain use cases

**Comment by joerdav** [comment]
Are you able to outline your use case a bit more? You mentioned you can't use templ due to the compilation step?

**Comment by gedw99** [comment]
Tanks for asking 

I can’t compile because my users are scientists and the whole platform is based on no code principles. 

At the moment I use a pipeline that is essentially like a vm to take the ast of the gui and bind the data and stream it out as xml. The xml is streamed out to renderers that render to Web, pdf, png and video. 

When an Ast gui  or data changes  the stream just pumps again so it htmx(ish), in that if your a user looking at a view it will refresh without you lifting a finger. I don’t use htmx,but instead nats. 

I was hoping to generate the html using templ. The ast would gen the xml and the Renderer would render the templ and this the html :)  but no dive I think .

All raater bits are generated and mixed into the html rendering pipelines, but. Think I will have to just write a ast to html Renderer.

I have to do the same for email too

**Comment by a-h** [comment]
I'm not sure about your use case, but I'd guess you've defined a standard-ish XML format for the data (or you could).

You'd define a struct that matches the XML structure and unmarshal the XML into it.

```go
package main

import (
	"encoding/xml"
	"net/http"

	"github.com/a-h/templ"
)

type Data struct {
	Values []TimeValue
}

type TimeValue struct {
	Time  string  `json:"time"`
	Value float64 `json:"value"`
}

const xmlData = `<Data>
 <Values>
  <Time>2019-04-11</Time>
  <Value>80.01</Value>
 </Values>
 <Values>
  <Time>2019-04-12</Time>
  <Value>96.63</Value>
 </Values>
 <Values>
  <Time>2019-04-13</Time>
  <Value>76.64</Value>
 </Values>
 <Values>
  <Time>2019-04-14</Time>
  <Value>81.89</Value>
 </Values>
 <Values>
  <Time>2019-04-15</Time>
  <Value>74.43</Value>
 </Values>
 <Values>
  <Time>2019-04-16</Time>
  <Value>80.01</Value>
 </Values>
 <Values>
  <Time>2019-04-17</Time>
  <Value>96.63</Value>
 </Values>
 <Values>
  <Time>2019-04-18</Time>
  <Value>76.64</Value>
 </Values>
 <Values>
  <Time>2019-04-19</Time>
  <Value>81.89</Value>
 </Values>
 <Values>
  <Time>2019-04-20</Time>
  <Value>74.43</Value>
 </Values>
</Data>
`

func main() {
	http.HandleFunc("/", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		// You might get the XML from a database, or from a HTTP form POST etc.
		var data Data
		err := xml.Unmarshal([]byte(xmlData), &data)
		if err != nil {
			panic(err.Error())
		}

		// Render the component and serve the results.
		htmlComponent := DataTable(data)
		httpHandler := templ.Handler(htmlComponent)
		httpHandler.ServeHTTP(w, r)
	}))
	http.ListenAndServe(":9090", nil)
}
```

Since you now have a type that defines the the data model, you can use a template.

```templ
package main

import "fmt"

templ DataTable(data Data) {
	<table>
		<tr>
			<th>Time</th>
			<th>Value</th>
		</tr>
		for _, v := range data.Values {
			<tr>
				<td>{ v.Time }</td>
				<td>{ fmt.Sprintf("%f.2", v.Value) }</td>
			</tr>
		}
	</table>
}
```

If you don't have a standard structure for the XML, then you could parse the XML and pull out "well known" XML elements by their name, unmarshal that element into its appropriate struct, and use appropriate sub-templates.

For example, if you have a `<StringItem>Value</StringItem>` type, you might pull that element out, use a template to render that element, then move on to the next XML element in the stream.

You might also have a `<KeyValue><Key>String</Key><Value>Value</Value></KeyValue>` - same thing applies.

Basically, you'd need to write some code that would take arbitrary input XML, and look through it for XML elements of specific names, unmarshal the element contents into a predefined struct, and then look up the appropriate pre-compiled template function to use to render it. You could have multiple templates for rendering the same struct, which are loaded dynamically based on your scientist's config.

Since templ outputs Go code, there's no side-stepping the compilation step. These days, it's possible to compile and run stuff in a container or Firecracker VM to provide a level of process isolation and compute control if you want to let people compile and run arbitrary code in your system.

**Comment by gedw99** [comment]
Thanks for the effort of that code. It makes sense.

MBit I have to either put in a Compiler as a Service OR just use traditional o templates.

min my case I am electing to use traditional go templates . 
Mir will still work well with htmx pattern

**Comment by a-h** [comment]
I think this discussion has concluded, so I'll close the issue. Feel free to reopen though.
